In [78]:
import pandas as pd

In [79]:
# read the dataset
df = pd.read_csv('../data/RTAQI.csv')
df.sample(5)

country           state         city  \
2460   India       Telangana    Hyderabad   
2535   India   Uttar_Pradesh    Moradabad   
59     India           Bihar  Muzaffarpur   
282    India  Madhya Pradesh     Jabalpur   
1105   India       Rajasthan      Bhiwadi   

                                            station          last_update  \
2460                     Kokapet, Hyderabad - TSPCB  14-04-2024 11:00:00   
2535              Kashiram Nagar, Moradabad - UPPCB  14-04-2024 11:00:00   
59    Muzaffarpur Collectorate, Muzaffarpur - BSPCB  14-04-2024 11:00:00   
282                      Gupteshwar, Jabalpur - JMC  14-04-2024 11:00:00   
1105          Vasundhara Nagar_UIT, Bhiwadi - RSPCB  14-04-2024 11:00:00   

       latitude  longitude pollutant_id  pollutant_min  pollutant_max  \
2460  17.393559  78.339194        PM2.5           51.0           65.0   
2535  28.849399  78.742362          SO2           35.0           44.0   
59    26.120900  85.364700           CO           42.0          102.0   
282   23.142888  79.916147           CO           20.0           42.0   
1105  28.207266  76.829265        OZONE           10.0          127.0   

      pollutant_avg  
2460           58.0  
2535           38.0  
59             50.0  
282            27.0  
1105           25.0

In [80]:
# check dataframe information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3233 entries, 0 to 3232
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   country        3233 non-null   object 
 1   state          3233 non-null   object 
 2   city           3233 non-null   object 
 3   station        3233 non-null   object 
 4   last_update    3233 non-null   object 
 5   latitude       3233 non-null   float64
 6   longitude      3233 non-null   float64
 7   pollutant_id   3233 non-null   object 
 8   pollutant_min  3030 non-null   float64
 9   pollutant_max  3030 non-null   float64
 10  pollutant_avg  3030 non-null   float64
dtypes: float64(5), object(6)
memory usage: 278.0+ KB


There appears to be missing values in column `pollutant_min`, `pollutant_max`, `pollutant_avg`

In [81]:
# checking dataframe statistics
df.describe()

latitude    longitude  pollutant_min  pollutant_max  pollutant_avg
count  3233.000000  3233.000000    3030.000000    3030.000000    3030.000000
mean     22.954409    78.541401      23.398680      81.019472      41.901650
std       5.325533     4.863931      25.261122      92.897985      43.427336
min       8.514909    70.909168       1.000000       1.000000       1.000000
25%      19.110740    75.565602       6.000000      19.000000      11.000000
50%      23.717634    77.287209      15.000000      50.000000      29.000000
75%      27.308328    80.327067      33.000000     103.000000      56.000000
max      31.620000    94.636574     306.000000     500.000000     340.000000

**Pollutant Min:**
The mean minimum pollutant value is approximately 23.40, indicating the average lowest pollutant concentration across observations. <br>
The minimum pollutant minimum value of 1 suggests the presence of outliers or very low pollutant levels in some observations. <br>
The maximum pollutant minimum value of 306 indicates relatively high minimum pollutant levels in some observations. <br>
The standard deviation of approximately 25.26 reflects the variability in minimum pollutant values across the dataset, showcasing fluctuating pollution levels. <br>

**Pollutant Max:**
The mean maximum pollutant value is approximately 81.02, indicating the average highest pollutant concentration across observations. <br>
The minimum pollutant maximum value of 1 suggests the presence of outliers or very low pollutant levels in some observations. <br>
The maximum pollutant maximum value of 500 indicates relatively high maximum pollutant levels in some observations. <br>
The standard deviation of approximately 92.90 reflects the variability in maximum pollutant values across the dataset, highlighting diverse pollution levels. <br>

**Pollutant Avg:**
The mean average pollutant value is approximately 41.90, indicating the average pollutant concentration across observations. <br>
The minimum pollutant average value of 1 suggests the presence of outliers or very low pollutant levels in some observations. <br>
The maximum pollutant average value of 340 indicates relatively high average pollutant levels in some observations. <br>
The standard deviation of approximately 43.43 reflects the variability in average pollutant values across the dataset, illustrating fluctuations in pollution levels over time or across locations. <br>

In [82]:
# checking for uniques in country
df.country.unique()

# since there is only one country in this dataset, we can drop this field
df.shape
df = df.drop('country', axis=1)
df.shape

(3233, 10)

In [83]:
# dropping latitude and longitude data not need it for this analysis
display(df.shape)
df=df.drop(['latitude', 'longitude'], axis=1)
display(df.shape)

(3233, 10)

(3233, 8)

In [84]:
# convert last_update to datetime
df['last_update'] = pd.to_datetime(df['last_update'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3233 entries, 0 to 3232
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   state          3233 non-null   object        
 1   city           3233 non-null   object        
 2   station        3233 non-null   object        
 3   last_update    3233 non-null   datetime64[ns]
 4   pollutant_id   3233 non-null   object        
 5   pollutant_min  3030 non-null   float64       
 6   pollutant_max  3030 non-null   float64       
 7   pollutant_avg  3030 non-null   float64       
dtypes: datetime64[ns](1), float64(3), object(4)
memory usage: 202.2+ KB


In [85]:
# glancing over the dataset
df.sample(5)

state       city                           station  \
2497  Uttar_Pradesh  Ghaziabad     Vasundhara, Ghaziabad - UPPCB   
1457        Gujarat  Ahmedabad         Rakhial, Ahmedabad - IITM   
584   Uttar_Pradesh  Firozabad     Nagla Bhau, Firozabad - UPPCB   
1118      Rajasthan  Dungarpur       Bhoiwada, Dungarpur - RSPCB   
500       Rajasthan    Karauli  Satyawati Vihar, Karauli - RSPCB   

             last_update pollutant_id  pollutant_min  pollutant_max  \
2497 2024-04-14 11:00:00          NO2           19.0           84.0   
1457 2024-04-14 11:00:00         PM10           50.0          115.0   
584  2024-04-14 11:00:00          NH3            2.0            2.0   
1118 2024-04-14 11:00:00        OZONE           31.0          103.0   
500  2024-04-14 11:00:00        PM2.5           17.0           88.0   

      pollutant_avg  
2497           47.0  
1457           86.0  
584             2.0  
1118           50.0  
500            41.0

In [86]:
# checking for missing values
df.isnull().sum()

state              0
city               0
station            0
last_update        0
pollutant_id       0
pollutant_min    203
pollutant_max    203
pollutant_avg    203
dtype: int64

We have 203 values missing for all three pollutant numerical values.

In [87]:
# checking for any pattern in missing values
df.sample(10)

state         city                                 station  \
740            Delhi        Delhi                  Aya Nagar, Delhi - IMD   
2469       Telangana    Hyderabad           Somajiguda, Hyderabad - TSPCB   
741            Delhi        Delhi                  Aya Nagar, Delhi - IMD   
2723         Haryana  Bahadurgarh         Arya Nagar, Bahadurgarh - HSPCB   
2834  Madhya Pradesh       Indore         Chhoti Gwaltoli, Indore - MPPCB   
3139   Uttar_Pradesh    Firozabad         Vibhab Nagar, Firozabad - UPPCB   
2195  Madhya Pradesh       Bhopal             Idgah Hills, Bhopal - MPPCB   
1575  Madhya Pradesh        Sagar         Deen Dayal Nagar, Sagar - MPPCB   
1020     Maharashtra         Pune               Mhada Colony, Pune - IITM   
2622           Bihar  Muzaffarpur  MIT-Daudpur Kothi, Muzaffarpur - BSPCB   

             last_update pollutant_id  pollutant_min  pollutant_max  \
740  2024-04-14 11:00:00         PM10           62.0          190.0   
2469 2024-04-14 11:00:00        PM2.5           18.0           18.0   
741  2024-04-14 11:00:00           CO           74.0          115.0   
2723 2024-04-14 11:00:00        PM2.5           40.0          117.0   
2834 2024-04-14 11:00:00         PM10           50.0          155.0   
3139 2024-04-14 11:00:00        PM2.5           54.0          121.0   
2195 2024-04-14 11:00:00        OZONE           47.0           84.0   
1575 2024-04-14 11:00:00           CO           47.0           98.0   
1020 2024-04-14 11:00:00        OZONE            1.0           29.0   
2622 2024-04-14 11:00:00           CO           55.0          108.0   

      pollutant_avg  
740           111.0  
2469           18.0  
741           113.0  
2723           65.0  
2834           91.0  
3139           72.0  
2195           54.0  
1575           62.0  
1020            3.0  
2622           55.0

When we have missing values, we have it for the entire row of data, likey due to sensor malfunctions, we can impute this data with mean to deal with them.

In [88]:
# checking for all the unique pollutants in the dataset
df.pollutant_id.unique()

array(['NH3', 'SO2', 'PM2.5', 'PM10', 'OZONE', 'CO', 'NO2'], dtype=object)

In [89]:
# List of all possible pollutants
possible_pollutants = ['NH3', 'SO2', 'PM2.5', 'PM10', 'OZONE', 'CO', 'NO2']

def impute_missing_values(df, possible_pollutants):
    # Group by 'pollutant_id' and calculate the mean pollutant values for each group
    pollutant_means = df.groupby('pollutant_id')[['pollutant_min', 'pollutant_avg', 'pollutant_max']].mean().to_dict(orient='index')
    
    # Function to impute missing values based on pollutant category means
    def impute_missing(row):
        pollutant_id = row['pollutant_id']
        if pd.isnull(row['pollutant_min']):
            row['pollutant_min'] = pollutant_means.get(pollutant_id, {}).get('pollutant_min', None)
        if pd.isnull(row['pollutant_avg']):
            row['pollutant_avg'] = pollutant_means.get(pollutant_id, {}).get('pollutant_avg', None)
        if pd.isnull(row['pollutant_max']):
            row['pollutant_max'] = pollutant_means.get(pollutant_id, {}).get('pollutant_max', None)
        return row
    
    # Apply the imputation function to each row of the DataFrame
    df_imputed = df.apply(impute_missing, axis=1)
    
    return df_imputed

In [90]:
df_imputed = impute_missing_values(df, possible_pollutants)

In [91]:
df.describe()

last_update  pollutant_min  pollutant_max  \
count                           3233    3030.000000    3030.000000   
mean   2024-04-14 10:59:59.999999744      23.398680      81.019472   
min              2024-04-14 11:00:00       1.000000       1.000000   
25%              2024-04-14 11:00:00       6.000000      19.000000   
50%              2024-04-14 11:00:00      15.000000      50.000000   
75%              2024-04-14 11:00:00      33.000000     103.000000   
max              2024-04-14 11:00:00     306.000000     500.000000   
std                              NaN      25.261122      92.897985   

       pollutant_avg  
count    3030.000000  
mean       41.901650  
min         1.000000  
25%        11.000000  
50%        29.000000  
75%        56.000000  
max       340.000000  
std        43.427336

In [92]:
df_imputed.describe()

last_update  pollutant_min  pollutant_max  \
count                           3233    3233.000000    3233.000000   
mean   2024-04-14 10:59:59.999999744      23.538871      81.495185   
min              2024-04-14 11:00:00       1.000000       1.000000   
25%              2024-04-14 11:00:00       6.000000      20.000000   
50%              2024-04-14 11:00:00      15.965909      51.000000   
75%              2024-04-14 11:00:00      34.000000     105.000000   
max              2024-04-14 11:00:00     306.000000     500.000000   
std                              NaN      24.871403      91.543673   

       pollutant_avg  
count    3233.000000  
mean       42.171106  
min         1.000000  
25%        12.000000  
50%        29.000000  
75%        58.000000  
max       340.000000  
std        42.921555

We have imputed all the missing values without making a significant difference in the central tendencies

In [94]:
# exporting the data
df_imputed.to_csv("../data/RTAQI_update.csv", index=False)